File: Modules/BuiltInAlgorithms/Sources/BinaryMergeSort.swift
import AlgorithmKit
import SortEngineKit

public struct BinaryMergeSort: SortAlgorithm {
    public let id = AlgorithmID(rawValue: "binarymergesort")
    public let metadata = AlgorithmMetadata(
        displayName: "Binary Merge Sort",
        category: .hybrid,
        sizeRange: 16...256,
        stable: true,
        timeComplexity: ComplexityBounds(best: "O(n log n)", average: "O(n log n)", worst: "O(n log n)"),
        spaceComplexity: "O(n)",
        iconName: "arrow.triangle.merge"
    )

    public init() {}

    public func record(into engine: inout RecordingEngine) {
        let n = engine.count
        guard n >= 2 else { return }

        let threshold = 32
        let tempHandle = engine.createAuxArray(length: n)

        func insertionSort(_ start: Int, _ end: Int) {
            var i = start + 1
            while i < end {
                var j = i
                while j > start && !engine.compare(j, j - 1) {
                    engine.swap(j - 1, j)
                    j -= 1
                }
                i += 1
            }
        }

        func merge(_ start: Int, _ mid: Int, _ end: Int) {
            var low = start
            var high = mid
            var merged: [Int] = []
            while low < mid && high < end {
                if engine.compare(high, low) {
                    merged.append(engine.values[low])
                    low += 1
                } else {
                    merged.append(engine.values[high])
                    high += 1
                }
            }
            while low < mid {
                merged.append(engine.values[low])
                low += 1
            }
            while high < end {
                merged.append(engine.values[high])
                high += 1
            }
            for i in 0..<merged.count {
                engine.writeAux(tempHandle, at: start + i, value: merged[i])
            }
            for i in 0..<merged.count {
                engine.setValue(start + i, merged[i])
            }
        }

        func mergeSort(_ start: Int, _ end: Int) {
            if end - start <= threshold {
                insertionSort(start, end)
                return
            }
            let mid = start + (end - start) / 2
            mergeSort(start, mid)
            mergeSort(mid, end)
            merge(start, mid, end)
        }

        mergeSort(0, n)

        engine.deleteAuxArray(tempHandle)
    }
}

File: App/Resources/AlgorithmDetails/binarymergesort/binarymergesort.java
import java.util.Arrays;

public class binarymergesort {
  private static final int THRESHOLD = 32;

  public static void insertionSort(int[] arr, int start, int end) {
    for (int i = start + 1; i < end; i++) {
      int j = i;
      while (j > start && arr[j] < arr[j - 1]) {
        int temp = arr[j - 1];
        arr[j - 1] = arr[j];
        arr[j] = temp;
        j--;
      }
    }
  }

  public static void merge(int[] arr, int start, int mid, int end) {
    int low = start;
    int high = mid;
    int[] merged = new int[end - start];
    int k = 0;
    while (low < mid && high < end) {
      if (arr[high] < arr[low]) {
        merged[k++] = arr[high++];
      } else {
        merged[k++] = arr[low++];
      }
    }
    while (low < mid) {
      merged[k++] = arr[low++];
    }
    while (high < end) {
      merged[k++] = arr[high++];
    }
    for (int i = 0; i < k; i++) {
      arr[start + i] = merged[i];
    }
  }

  public static void mergeSort(int[] arr, int start, int end) {
    if (end - start <= THRESHOLD) {
      insertionSort(arr, start, end);
      return;
    }
    int mid = start + (end - start) / 2;
    mergeSort(arr, start, mid);
    mergeSort(arr, mid, end);
    merge(arr, start, mid, end);
  }

  public static void sort(int[] arr) {
    int n = arr.length;
    if (n < 2) {
      return;
    }
    mergeSort(arr, 0, n);
  }

  public static void main(String[] args) {
    int[] array = new int[] {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};
    sort(array);
    System.out.println(Arrays.toString(array));
  }
}
